numpy/array_api/_typing.py

```diff
@@ -33,12 +35,26 @@
 
 # This should really be recursive, but that isn't supported yet. See the
 # similar comment in numpy/typing/_array_like.py
-NestedSequence = Sequence[Sequence[Any]]
+_T = TypeVar("_T")
+NestedSequence = Sequence[Sequence[_T]]
 
 Device = Literal["cpu"]
-Dtype = Type[
-    Union[int8, int16, int32, int64, uint8, uint16, uint32, uint64, float32, float64]
-]
+if TYPE_CHECKING or sys.version_info >= (3, 9):
+    Dtype = dtype[Union[
+        int8,
+        int16,
+        int32,
+        int64,
+        uint8,
+        uint16,
+        uint32,
+        uint64,
+        float32,
+        float64,
+    ]]
+else:
+    Dtype = dtype
```

Comment:
Previous `Dtype` was defined based on a union of concrete `np.dtype` instance, _e.g._ `Type[Union[np.dtype(np.int8), ...]]`. 

As far as I'm aware there is not a single type checker that supports function-calls inside an annotation (too dynamic, I presume) and even during runtime this expression will raise. As a solution, `Dtype`' has now been changed to a more conventional subscribed type.

Reply:
The dtypes in the annotation were defined just as the names, like `int8`, not a complex expression like `np.dtype(np.int8)`. Are type checkers really not able to work even with basic names. Any function that has Dtype should basically only pass a type check if it uses the exact same name from this namespace, like `dtype=int8`.

Reply:
> The dtypes in the annotation were defined just as the names, like `int8`, not a complex expression like `np.dtype(np.int8)`

An annotation that is purely defined based on a name (ignoring any info about the type) could hypothetically work if we ever get something like `Literal` support for arbitrary singletons. Until then, type checkers face the challenge of identifying whether or not the assignments below is intended to represent a normal dtype instances (_e.g._ `np.dtype(np.int64)`) or if they are dtype type-aliases (_e.g._ `np.dtype[np.int64]`). The way they deal with this challenge is to straight up disallow the use of callable outputs as type aliases.

https://github.com/numpy/numpy/blob/a136dba9d52d77691380d3b5385e090e5b7d7e6b/numpy/array_api/_dtypes.py#L5-L7